components/script/dom/documentorshadowroot.rs

```diff
@@ -0,0 +1,323 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```

Comment:
I'm assuming most of this file is moving code, so I'm not reviewing that code in much detail. Hopefully that's fine.